compress.go

```diff
@@ -0,0 +1,244 @@
+	// true length is contained in comprLength
+	if uncompressedLength == 0 {
+		cr.bytesBuf = append(cr.bytesBuf, comprData...)
+		return nil
+	}
+
+	// write comprData to a bytes.buffer, then read it using zlib into data
+	br := bytes.NewReader(comprData)
+
+	resetter, ok := cr.zr.(zlib.Resetter)
+
+	if ok {
+		err := resetter.Reset(br, []byte{})
+		if err != nil {
+			return err
+		}
+	} else {
+		cr.zr, err = zlib.NewReader(br)
+		if err != nil {
+			return err
+		}
+	}
+
+	defer cr.zr.Close()
+
+	//use existing capacity in bytesBuf if possible
+	offset := len(cr.bytesBuf)
+	if cap(cr.bytesBuf)-offset < uncompressedLength {
+		old := cr.bytesBuf
+		cr.bytesBuf = make([]byte, offset, offset+uncompressedLength)
+		copy(cr.bytesBuf, old)
+	}
+
+	data := cr.bytesBuf[offset : offset+uncompressedLength]
+
+	lenRead := 0
+
+	// http://grokbase.com/t/gg/golang-nuts/146y9ppn6b/go-nuts-stream-compression-with-compress-flate
+	for lenRead < uncompressedLength {
+		n, err := cr.zr.Read(data[lenRead:])
+		lenRead += n
+
+		if err == io.EOF {
+			if lenRead < uncompressedLength {
+				return io.ErrUnexpectedEOF
+			}
+			break
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	cr.bytesBuf = append(cr.bytesBuf, data...)
+
+	return nil
+}
+
+func (cw *compressedWriter) Write(data []byte) (int, error) {
+	// when asked to write an empty packet, do nothing
+	if len(data) == 0 {
+		return 0, nil
+	}
+	totalBytes := len(data)
+
+	length := len(data) - 4
+
+	maxPayloadLength := maxPacketSize - 4
+
+	for length >= maxPayloadLength {
+		// cut off a slice of size max payload length
+		dataSmall := data[:maxPayloadLength]
+		lenSmall := len(dataSmall)
+
+		var b bytes.Buffer
+		writer := zlib.NewWriter(&b)
+		_, err := writer.Write(dataSmall)
+		writer.Close()
+		if err != nil {
+			return 0, err
+		}
+
+		// if compression expands the payload, do not compress
+		useData := b.Bytes()
+
+		if len(useData) > len(dataSmall) {
+			useData = dataSmall
+			lenSmall = 0
+		}
+
+		err = cw.writeComprPacketToNetwork(useData, lenSmall)
+		if err != nil {
+			return 0, err
+		}
+
+		length -= maxPayloadLength
+		data = data[maxPayloadLength:]
+	}
+
+	lenSmall := len(data)
+
+	// do not attempt compression if packet is too small
+	if lenSmall < minCompressLength {
+		err := cw.writeComprPacketToNetwork(data, 0)
+		if err != nil {
+			return 0, err
+		}
+
+		return totalBytes, nil
+	}
+
+	var b bytes.Buffer
+	writer := zlib.NewWriter(&b)
+
+	_, err := writer.Write(data)
+	writer.Close()
+
+	if err != nil {
+		return 0, err
+	}
+
+	// if compression expands the payload, do not compress
+	useData := b.Bytes()
+
+	if len(useData) > len(data) {
+		useData = data
+		lenSmall = 0
+	}
+
+	err = cw.writeComprPacketToNetwork(useData, lenSmall)
+
+	if err != nil {
+		return 0, err
+	}
+	return totalBytes, nil
+}
+
+func (cw *compressedWriter) writeComprPacketToNetwork(data []byte, uncomprLength int) error {
+	data = append(cw.header, data...)
```

Comment:
To avoid allocation, you can require 7byte space is reserved in payload to caller.
Caller can reserve it like:

```go
cHeader := make([]byte, 7)
var cBuff = new(bytes.Buffer)
zw := zlib.NewWriter(cBuff)
...
cBuff.Reset()
cBuff.Write(cHeader)
zw.Reset(cBuff)
...
```